File: web/src/entities/dialogue.ts
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { GameState } from '../gameState.js';
import { GlobalOptions } from '../options.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';
import type { BitmapFont } from '../media/font.js';

const enum CharacterPortrait { DoctorGreen, Ted }

interface DialogueLine {
  readonly text: string;
  readonly portrait: CharacterPortrait;
}

export class Dialogue extends Entity {
  private readonly background: Animation;
  private currentCharacter = 0;
  private currentLine = 0;
  private readonly doctorGreenPortrait: Animation;
  private done = false;
  private readonly file: string;
  private readonly font: BitmapFont;
  private frameCounter = 0;
  private readonly lines: DialogueLine[] = [];
  private running = false;
  private runWithoutHero = false;
  private readonly tedPortrait: Animation;

  constructor(filename: string) {
    super();
    this.file = filename;
    this.background = Resource.getAnimation('data/images/dialogue.bmp', 1);
    this.doctorGreenPortrait = Resource.getAnimation('data/images/doctor_green_portrait.bmp', 1);
    this.tedPortrait = Resource.getAnimation('data/images/ted_portrait.bmp', 1);
    this.font = Resource.getFont('data/images/font.bmp');

    const text = Resource.getText(filename);
    const rawLines = text.split('\n');
    for (const line of rawLines) {
      if (line.length === 0) continue;
      const ch = line.charAt(0);
      const portrait = ch === 'D' ? CharacterPortrait.DoctorGreen : CharacterPortrait.Ted;
      this.lines.push({ text: line.substring(1), portrait });
    }

    this.setSize(new ImmutableVec2(10, 10 * 200));
  }

  getLayer(): number { return 3; }

  setRunWithoutHero(): void {
    this.runWithoutHero = true;
    this.running = true;
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    if (this.done || !this.running) return;

    let position = 240 - this.background.getFrameHeight();
    if (GlobalOptions.dialogueAtTop()) position = 10;

    this.background.drawFrame(ctx, 0, 0, position, false, GlobalOptions.dialogueAtTop());

    const line = this.lines[this.currentLine];
    if (!line) return;

    const pX = 4;
    const pY = position + this.background.getFrameHeight() - this.tedPortrait.getFrameHeight() - 3;

    if (line.portrait === CharacterPortrait.Ted)
      this.tedPortrait.drawFrame(ctx, 0, pX, pY);
    if (line.portrait === CharacterPortrait.DoctorGreen)
      this.doctorGreenPortrait.drawFrame(ctx, 0, pX, pY);

    let fontPosition = position + this.background.getFrameHeight() - 16;
    if (GlobalOptions.dialogueAtTop()) fontPosition = position + 5;

    this.font.drawWrap(ctx, line.text, 55, fontPosition, 300, this.currentCharacter);
  }

  override update(): void {
    if (this.done) return;

    if (!this.runWithoutHero) {
      if (this.myRoom.getHero().Collides(this)) {
        this.running = true;
      }
    }

    if (!this.running) return;

    this.frameCounter++;

    const line = this.lines[this.currentLine];
    if (!line) return;

    if (this.frameCounter > 2 && this.currentCharacter < line.text.length) {
      this.currentCharacter++;
      this.frameCounter = 0;
      Sound.playSample('data/sounds/beep');
    }

    if (this.frameCounter > 90 && this.lines.length > this.currentLine) {
      this.currentLine++;
      this.currentCharacter = 0;
      this.frameCounter = 0;

      if (this.lines.length === this.currentLine) {
        const key = 'dialogue_' + this.file;
        GameState.put(key, 1);
        this.done = true;
        if (!this.runWithoutHero) this.remove();
      }
    }
  }
}

File: wii/src/entities/dialogue.cpp
#include "entities/dialogue.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "media/font.h"
#include "room.h"
#include "game_state.h"
#include "constants.h"
#include <sstream>

Dialogue::Dialogue(const std::string& filename)
    : file_(filename) {
    background_ = Resource::getAnimation("data/images/dialogue.bmp", 1);
    doctorGreenPortrait_ = Resource::getAnimation("data/images/doctor_green_portrait.bmp", 1);
    tedPortrait_ = Resource::getAnimation("data/images/ted_portrait.bmp", 1);
    font_ = Resource::getFont("data/images/font.bmp");

    std::string text = Resource::getText(filename);
    std::istringstream stream(text);
    std::string line;
    while (std::getline(stream, line)) {
        if (line.empty()) continue;
        char ch = line[0];
        CharacterPortrait portrait = (ch == 'D') ? CharacterPortrait::DoctorGreen : CharacterPortrait::Ted;
        lines_.push_back({line.substr(1), portrait});
    }

    setSize(Vec2(10, 10 * 200));
}

void Dialogue::setRunWithoutHero() {
    runWithoutHero_ = true;
    running_ = true;
}

void Dialogue::draw(SDL_Renderer* renderer, int /*offsetX*/, int /*offsetY*/, int /*layer*/) {
    if (done_ || !running_) return;

    int position = SCREEN_HEIGHT - background_->getFrameHeight();

    background_->drawFrame(renderer, 0, 0, position);

    if (currentLine_ >= static_cast<int>(lines_.size())) return;
    const DialogueLine& line = lines_[currentLine_];

    int pX = 4;
    int pY = position + background_->getFrameHeight() - tedPortrait_->getFrameHeight() - 3;

    if (line.portrait == CharacterPortrait::Ted)
        tedPortrait_->drawFrame(renderer, 0, pX, pY);
    if (line.portrait == CharacterPortrait::DoctorGreen)
        doctorGreenPortrait_->drawFrame(renderer, 0, pX, pY);

    int fontPosition = position + background_->getFrameHeight() - 16;

    font_->drawWrap(renderer, line.text, 55, fontPosition, 300, currentCharacter_);
}

void Dialogue::update() {
    if (done_) return;

    if (!runWithoutHero_) {
        if (room_->getHero()->Collides(*this)) {
            running_ = true;
        }
    }

    if (!running_) return;

    dialogueFrameCounter_++;

    if (currentLine_ >= static_cast<int>(lines_.size())) return;
    const DialogueLine& line = lines_[currentLine_];

    if (dialogueFrameCounter_ > 2 && currentCharacter_ < static_cast<int>(line.text.length())) {
        currentCharacter_++;
        dialogueFrameCounter_ = 0;
        Sound::playSample("data/sounds/beep");
    }

    if (dialogueFrameCounter_ > 90 && static_cast<int>(lines_.size()) > currentLine_) {
        currentLine_++;
        currentCharacter_ = 0;
        dialogueFrameCounter_ = 0;

        if (static_cast<int>(lines_.size()) == currentLine_) {
            std::string key = "dialogue_" + file_;
            GameState::put(key, 1);
            done_ = true;
            if (!runWithoutHero_) remove();
        }
    }
}
